### PR TITLE
Added metrics calculation on scala UDFs

### PR DESCRIPTION
**Pull request by netang**
* Added UDFs on scala to calculate metrics.
* Added tests to check UDFs on scala.

To build jar file from scala code you can use the following commands:
```bash
cd scala
sbt package
```

To use the calculation of metrics on scala UDFs, you need to create a metric instance with the `use_scala_udf=True` argument.  For example,
```
NDCG(use_scala_udf=True)
```
You also need to include the jar in the application.

### CODE REVIEW

File: tests/test_metrics.py [hunk 1]
```diff
@@ -180,12 +180,20 @@ def test_pred_is_bigger(quality_metrics, one_user, two_users):
     "gt_users, result",
     [(False, {3: 2 / 3, 1: 1 / 3}), (True, {3: 1 / 4, 1: 0 / 3})],
 )
-def test_hit_rate_at_k(recs, true, true_users, gt_users, result):
-    users = true_users if gt_users else None
-    assertDictAlmostEqual(
-        HitRate()(recs, true, [3, 1], users),
-        result,
-    )
+class TestHitRate:
```
Comment (monkey0head) [inline tests/test_metrics.py:183]: we have not used test classes in RePlay before. how do they help here? 
why do you duplicate tests if you can parametrize existing tests with use_scala_udf param?

Comment (netang) [reply tests/test_metrics.py:183]: > why do you duplicate tests if you can parametrize existing tests with use_scala_udf param?

I will update PR with additional test param, thx

File: replay/metrics/base_metric.py [hunk 2]
```diff
@@ -254,6 +268,16 @@ def _get_metric_distribution(self, recs: DataFrame, k: int) -> DataFrame:
         :param k: depth cut-off
         :return: metric distribution for different cut-offs and users
         """
+        if self._use_scala_udf:
+            # Possibly bad approach to define column names for udf call
```
Comment (monkey0head) [inline replay/metrics/base_metric.py:272]: we believe that `recs` dataframe comes from get_enriched_recommendations and has pre-defined structure. better add an order of columns to the docstring in :param recs: recommendations

Comment (monkey0head) [reply replay/metrics/base_metric.py:272]: so cols could be columns[1:]

**Review by monkey0head — APPROVED** [review]

**Comment by monkey0head** [comment]
merged as https://github.com/sb-ai-lab/RePlay/pull/36

**Pull request closed without merging**